keymanapp/keyman
# fix(web): support variable stores with predictive text

Fixes #2924.

When a keyboard includes variable stores, predictive text fat-fingering tests will no longer affect the variable store values.

The failure mode can be seen by using the test keyboard and model shared at https://community.software.sil.org/t/lexical-models-break-variable-stores-at-least-for-mobile-keyboards/3415/3 with a released version of KeymanWeb.

This makes changes to both KeymanWeb and Keyman Developer. Note that the majority of files are just source files for the unit test (under web/testing/issue2924/source), which you do not need to review.

# User Testing

For these tests, run them under Chrome mobile emulation or on a mobile device/mobile device emulator.

* **TEST_BASELINE:** Test predictive text for errors

  Load a test keyboard that includes predictive text, e.g. Prediction UI Testing in the Testing for KeymanWeb, and use it. Use Developer Tools to ensure no error messages are raised.

* **TEST_OLD_FAT_FINGER:** Verify that the old keyboard still goes wrong

  1. With the manual test page "Test variable stores and predictive text (2924)", select the "oldtestvariable" keyboard (under English)
  2. Press Set, then Reveal 4 times. You will see `(+)` followed by either `+` or `-` four times.
  3. Press Clear, then Reveal 4 times. You should see `(-)`, followed by either `+` or `-` four times.

  The behaviour will not appear consistent -- this demonstrates the bug. There is no need to fail this test :grin:

* **TEST_FAT_FINGER:** Verify that predictive text no longer triggers changes to variable stores

  1. With the manual test page "Test variable stores and predictive text (2924)", select the "testvariable" keyboard (under English)
  2. Press Set, then Reveal 4 times. You should see `(+)++++`.
  3. Delete the text, then press Clear, then Reveal 4 times. You should see `(-)----`.
